docs: Add API and CLI documentation (#3917)
dagger/dagger: docs/current/api/975146-concepts.md

---
slug: /api/concepts
---

> # Key Concepts

Author: I struggled a bit with how best to structure this tutorial. I decided finally on this concepts-based approach, as it seems to fit better for an API than the step-wise sequential walkthrough we have for the SDKs. I'd like to have feedback on whether this approach is OK or if I should consider a different structure altogether.

## Introduction

> One of Dagger's most powerful features is cross-language instrumentation: teams can use each other's tools without learning each other's language. Dagger achieves this by exposing its functionality through GraphQL, which serves as a low-level language-agnostic API framework for all Dagger operations.

Reviewer: ```suggestion
One of Dagger's key aspects is cross-language instrumentation: teams can leverage their familiar languages  without the need to learn a new SDL to write their CI/CD pipelines . Dagger allows this by exposing its functionality through GraphQL which serves as a low-level language-agnostic API framework for all Dagger operations.
```

This approach has two main benefits:

- It enables CI/CD pipeline authors to use whichever language feels natural and intuitive to them, without having to think about interoperability.

> - Rather than reimplementing the same automation in every possible programming language, it only needs to be implemented once and is then reusable by all. As a result, two cooperating teams no longer necessarily have to come to consensus on a common language.

Reviewer 2: This is very well-put, but sadly it will not be true until we ship extensions.

However, it does provide *some* interoperability benefits.

* Common runtime environment. Even if you write pipelines in different languages, they are all run by the same engine. So logging, scaling, monitoring, etc. only need to be solved once for everyone.
* Common low-level language. Even if teams can't directly re-use each other's pipelines (yet) they can take advantage of the fact that they all produce the same GraphQL queries. Similar to how developing for the same platform, even in different languages, creates opportunities for knowledge sharing and collaboration.

Author: Added

This tutorial explains the key concepts you need to use the Dagger GraphQL API.

:::note
This tutorial explains some basic GraphQL concepts in the context of the Dagger GraphQL API, but it is not intended as a full-featured GraphQL tutorial. To learn more about GraphQL and how it works, read the [GraphQL tutorial](https://graphql.org/learn/).
:::

## Objects, Fields and Build Operations

Let's start with a basic query:

```graphql
query {
  container {
    from(address: "alpine:latest") {
      exec(args: ["apk", "info"]) {

>         stdout {
>           contents
>         }

Reviewer 2: ```suggestion
        stdout
```

As of today's release the API for `stdout` has changed.

Author: Fixed

>       exec(args: ["apk", "info"]) {

Reviewer: ```suggestion
      withExec(args: ["apk", "info"]) {
```

    }
  }
}
```

This query represents a very simple Dagger pipeline. In plain English, it instructs Dagger to "download the latest `alpine` container image, run the command `apk info` in that image, and print the results of the command to the standard output device".

Here's an example of the command output - a list of the packages installed in the image:

```graphql
{
 "data": {
  "container": {
   "from": {

>     "exec": {

Reviewer: ```suggestion
    "withExec": {
```

    "exec": {
     "stdout": {
      "contents": "alpine-baselayout-data\nmusl\nbusybox\nalpine-baselayout\nalpine-keys\nca-certificates-bundle\nlibcrypto1.1\nlibssl1.1\nssl_client\nzlib\napk-tools\nscanelf\nmusl-utils\nlibc-utils\n"
     }
    }
   }
  }
 }
}
```

A GraphQL schema works by defining one or more object types, and then fields (and field arguments) on those object types. The fields of an object type can themselves be objects, allowing for different entities to be logically connected with each other. API users can then perform GraphQL queries to return all or some fields of an object.

> 

Reviewer: > * It's almost comparable to a bash pipeline where you pass data along and each step manipulates it in some way

by bash pipeline you mean piping multiple commands? I guess that analogy is correct, since basically it's a pipe of transformations to the initial state. Isn't "reactive programming" (https://en.wikipedia.org/wiki/Reactive_programming) the closest term?

> 

Author: I've incorporated some of @sipsma 's comments in the draft now. Thanks.

1. Consider the first level of the previous query:

  ```graphql
  query {
    container {
      from(address: "alpine:latest") {
        ...
      }
    }
  }
  ```

  This query requests the `from` field of Dagger's `Container` object type, passing it the address of a container image as argument. To resolve this, Dagger will initialize a container from the image published at the given address and return a `Container` object representing the new container image.

1. Now, consider the next level of the query:

  ```graphql
  query {
    container {
      from(address: "alpine:latest") {

>         exec(args: ["apk", "info"]) {

Reviewer: ```suggestion
        withExec(args: ["apk", "info"]) {
```

        exec(args: ["apk", "info"]) {
          ...
        }
      }
    }
  }
  ```

  Here, the query requests the `exec` field of the `Container` object returned in the previous step, passing it the command to be executed as an array of arguments. To resolve this, Dagger will  define the command for execution in the container image and return a `Container` object containing the execution results.

1. Finally, consider the innermost level of the query:

  ```graphql
  query {
    container {
      from(address: "alpine:latest") {
        exec(args: ["apk", "info"]) {

>           stdout {
>             contents
>           }

Reviewer 2: ```suggestion
          stdout
```

Author: Fixed

>         exec(args: ["apk", "info"]) {

Reviewer: ```suggestion
        withExec(args: ["apk", "info"]) {
```

        }
      }
    }
  }
  ```

  Here, the query requests the `stdout` field of the `Container` object returned in the previous step. To resolve this, Dagger will return a `File` object type containing the result of the last executed command. The output of the command is retrieved by addressing the `contents` field of the `File` type, which returns a `String`.

> ## Lazy-Loading

Author: I would like feedback on which concepts should be discussed in this guide. The ones I've planned are lazy-loading and chaining, but maybe there are other (better?) ones?

> Dagger executes queries "lazily". This means that the individual build operations are performed only when a leaf value (scalar or enum) is requested. In the previous example, Dagger executes the query only when the `stdout` field is requested. This is a key advantage of Dagger: it allows users to write queries (or code, if using an SDK) in a procedural manner, then executes those queries in parallel for maximum speed.

Author: I read through https://github.com/dagger/dagger/issues/3555, and I understand this might change. I'm assuming it's correct for now, but let me know if not. Also, is there a way for a user to identify in the API reference which fields trigger execution? If not, can you please provide a short list of common ones here?

Reviewer 2: There isn't a list of fields that trigger execution. But it would be useful to define a simple rule that people can memorize.

The rule I've been using so far is something like:

*The Dagger API is data-oriented. You query for data, and dagger returns that data. If executing a command is necessary to return that data, it will do so. When crafting a pipeline, ask yourself: what data do you need that pipeline to return? Is it an artifact? The result or standard output of a command? Most of the time, actually executing a command is just an implementation detail of what you actually want*

> Dagger executes queries "lazily". This means that the individual build operations are performed only when a leaf value (scalar or enum) is requested. In the previous example, Dagger executes the query only when the `stdout` field is requested. This is a key advantage of Dagger: it allows users to write queries (or code, if using an SDK) in a procedural manner, then executes those queries in parallel for maximum speed.

Author: @sipsma added more detail in this section. Please review and add feedback.

> [example TBD]

Reviewer: We've synced with Vikram today and we have a really nice example to showcase this :hugs:

> [example TBD]

Author: Added this now

1. Now, consider the next level of the query:

  ```graphql
  query {
    container {
      from(address: "alpine:latest") {
        withExec(args: ["apk", "info"]) {
          ...
        }
      }
    }
  }
  ```

  Here, the query requests the `withExec` field of the `Container` object returned in the previous step, passing it the command to be executed as an array of arguments. To resolve this, Dagger will  define the command for execution in the container image and return a `Container` object containing the execution results.

1. Finally, consider the innermost level of the query:

  ```graphql
  query {
    container {
      from(address: "alpine:latest") {
        withExec(args: ["apk", "info"]) {
          stdout
        }
      }
    }
  }
  ```

  Here, the query requests the `stdout` field of the `Container` object returned in the previous step. To resolve this, Dagger will return a `String` containing the result of the last executed command.

## Object Identifiers

In a GraphQL schema, every object exposes an `id` field. This ID serves to uniquely identify and retrieve an object, and is also used by GraphQL's caching mechanism.

> In the Dagger GraphQL API too, objects expose an ID but here, the ID is actually a base64-encoded block representing the object's state at a given time. Objects like `Container` and `Directory` should be thought of as collections of state, which are updated by subsequent field resolutions and whose ID represents their state at the instant of field resolution.

Author: Fixed

The output of this query is a base64-encoded block representing the state of the container filesystem and Dagger's execution plan, similar to the output shown below:

```graphql
{
  "data": {
    "container": {
      "from": {
        "withExec": {
          "id": "eyJmc...9fQ=="
        }
      }
    }
  }
}
```

Now, execute a second query as follows, replacing the placeholder with the contents of the `id` field from the previous query:

```graphql
query {
  container(id: "YOUR-ID-HERE") {
    withExec(args: ["ls", "/tmp"]) {
      stdout
    }
  }
}
```

<ButtonExternal label="Try it!" url="https://play.dagger.cloud/playground/DKUH7PI5Yt0" />

This second query instructs Dagger to:

- initialize a container using the filesystem state provided in the ID
- run the `ls` command to list the files in the `/tmp/` directory (this resolves the `withExec` field)
- return the output of the command (this resolves the final `stdout` field)

This second query returns the following output:

```graphql
{
  "data": {
    "container": {
      "withExec": {
        "stdout": "myfile\n"
      }
    }
  }
}
```

> As this example demonstrates, Dagger object IDs hold the state of their corresponding object. This state can be transferred from one query to another, or even from one Dagger instance to another. It can also be updated or used as input to other objects.

Author: Fixed

## Lazy Evaluation

> GraphQL executes queries "lazily". This means that query resolution is triggered only when a leaf value (scalar) is requested. In the previous example, Dagger resolves the query and performs the build operations only when the `stdout` field is requested.

Author: Modified

  container {
    from(address: "alpine:latest") {
      withExec(args: ["apk", "add", "curl"]) {
        withExec(args: ["curl", "YOUR-WEBHOOK-URL-HERE"]) {
          stdout
        }
      }
    }
  }
}
```

<ButtonExternal label="Try it!" url="https://play.dagger.cloud/playground/SLtXQ4lvqNS" />

This time, Dagger both prepares and executes the plan, because that is the only way to resolve the `stdout` field. Check the Webhook.site dashboard and the HTTP request sent by the `curl` command will be visible in the request log, as shown below:

![Webhook.site request](/img/current/api/webhook-2.png)

This example demonstrates the lazy evaluation model in action. The client requests data, and the Dagger GraphQL API returns that data. If executing a command is necessary to return that data, it will do so; if not, it will not.

:::note
Lazy evaluation is a key advantage of GraphQL. It allows users to write queries (or code, if using an SDK) in a procedural manner, and resolves those queries only when necessary to return data to the user. Dagger leverages this lazy evaluation model to optimize and parallelize query execution for maximum speed and performance.
:::

## Conclusion

> This tutorial introduced you to the Dagger GraphQL API. It explained the reasons for using GraphQL and walked you through the components of a Dagger GraphQL API query. It also explained how Dagger leverages some of GraphQL's features, such as unique identifiers and lazy evaluation, to improve the portability and speed of CI/CD pipeline execution..

Author: Fixed
